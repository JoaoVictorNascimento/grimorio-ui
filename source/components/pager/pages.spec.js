import '../../../internals/test/helper';

import Pager from './pager-component';
import Select from '../select';

/** @test {Pager} */
describe('Pager component', () => {
/** @test {Pager#render} */
  describe('#render', () => {
    let wrapper;

    const props = {
      hasPerpage: true,
      onLimitChange: jest.fn(),
      onClickPagination: jest.fn(),
      hasFirstLast: false,
      hasPagination: false,
      count: 100,
      limit: 15,
      offset: 20,
      limitList: [10, 15, 50],
    }

    beforeEach(() => {
      wrapper = shallow(
        <Pager {...props} />
      );

      jest.spyOn(props, 'onLimitChange');
      jest.spyOn(props, 'onClickPagination');
    })
    it('render correctly', () => {
      expect(wrapper.debug()).toMatchSnapshot();
      expect(wrapper.find('.showing').text()).toBe(`${props.offset} - ${props.offset + props.limit} de ${props.count}`);
    });

     it('render correctly with hasFirstLast', () => {
      wrapper.setProps({ hasFirstLast: true });
      expect(wrapper.debug()).toMatchSnapshot();
    });
    it('onClick pagination first button', () => {
      wrapper.setProps({ hasFirstLast: true });
      wrapper.find('.nav .btn').at(0).simulate('click');
      expect(props.onClickPagination).toBeCalledWith('first');
    });

    it('onClick pagination prev button', () => {
      wrapper.setProps({ hasFirstLast: true });
      wrapper.find('.nav .btn').at(1).simulate('click');
      expect(props.onClickPagination).toBeCalledWith('prev');
    });

    it('onClick pagination next button', () => {
      wrapper.setProps({ hasFirstLast: true });
      wrapper.find('.nav .btn').at(2).simulate('click');
      expect(props.onClickPagination).toBeCalledWith('next');
    });

    it('onClick pagination last button', () => {
      wrapper.setProps({ hasFirstLast: true });
      wrapper.find('.nav .btn').at(3).simulate('click');
      expect(props.onClickPagination).toBeCalledWith('last');
    });

    it('render correctly with hasPagination', () => {
      wrapper.setProps({ hasPagination: true });
      expect(wrapper.debug()).toMatchSnapshot();
    });

    it('onClick pagination number button', () => {
      wrapper.setProps({ hasPagination: true });
      wrapper.find('.pagesWrap .btn').at(1).simulate('click');
      expect(props.onClickPagination).toBeCalledWith('number', 2);
    });

    it('first, prev, next and last navigators are disabled when pager has 1 page only', () => {
      const updated = {
        count: 8,
        limit: 10,
        offset: 0,
        hasFirstLast: true,
      };

      wrapper.setProps(updated);
      expect(wrapper.find('Button[name="first"]').prop('disabled')).toBe(true);
      expect(wrapper.find('Button[name="prev"]').prop('disabled')).toBe(true);
      expect(wrapper.find('Button[name="next"]').prop('disabled')).toBe(true);
      expect(wrapper.find('Button[name="last"]').prop('disabled')).toBe(true);
      expect(wrapper.find('.showing').text()).toBe(`${updated.offset} - ${updated.count} de ${updated.count}`);
    });

    it('only next and last navigators are disabled on the last page', () => {
      const updated = {
        count: 68,
        limit: 30,
        offset: 60,
        hasFirstLast: true,
      };
      wrapper.setProps(updated);
      expect(wrapper.find('Button[name="first"]').prop('disabled')).toBe(false);
      expect(wrapper.find('Button[name="prev"]').prop('disabled')).toBe(false);
      expect(wrapper.find('Button[name="next"]').prop('disabled')).toBe(true);
      expect(wrapper.find('Button[name="last"]').prop('disabled')).toBe(true);
      expect(wrapper.find('.showing').text()).toBe(`${updated.offset} - ${updated.count} de ${updated.count}`);
    });

    it('return null when count is undefined', () => {
      wrapper.setProps({ count: undefined });
      expect(wrapper.html()).toEqual(null);
    });

    it('return null when limit is undefined', () => {
      wrapper.setProps({ limit: undefined });
      expect(wrapper.html()).toEqual(null);
    });

    it('return null when offset is undefined', () => {
      wrapper.setProps({ offset: undefined });
      expect(wrapper.html()).toEqual(null);
    });

    it('should call function when changes perpage select', () => {
      const obj = { value: 10 };
      wrapper.find('.perpage').find(Select).prop('onSelect')(obj);
      expect(props.onLimitChange).toHaveBeenCalledWith(obj.value);
    });

    it('change start and end pages when currentPage <= 6', () => {
      wrapper.setProps({ offset: 60, hasPagination: true, count: 500 });
      expect(wrapper.debug()).toMatchSnapshot();
    });

    it('change start and end pages when currentPage + 4 >= maxPages', () => {
      wrapper.setProps({ offset: 300, hasPagination: true, count: 300 });
      expect(wrapper.debug()).toMatchSnapshot();
    });

    it('change start and end pages when is a default value', () => {
      wrapper.setProps({ offset: 100, hasPagination: true, count: 500 });
      expect(wrapper.debug()).toMatchSnapshot();
    });
  });
});
