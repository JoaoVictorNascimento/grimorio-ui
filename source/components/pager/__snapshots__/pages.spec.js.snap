// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager component #render change start and end pages when currentPage + 4 >= maxPages 1`] = `
<div
  className="holdPager"
>
  <div
    className="perpage"
  >
    <Select
      className="select"
      closeOnClickOutside={true}
      defaultValue={15}
      disabled={false}
      height="auto"
      isMobile={false}
      items={Array []}
      onSelect={[Function]}
      open={null}
      position="bottom"
      sortItems={false}
      type="select"
      value={false}
    >
      <SelectOption
        value={10}
      >
        10 por página
      </SelectOption>
      <SelectOption
        value={15}
      >
        15 por página
      </SelectOption>
      <SelectOption
        value={50}
      >
        50 por página
      </SelectOption>
    </Select>
  </div>
  <div
    className="showing"
  >
    300
     - 
    320
     de 
    300
  </div>
  <div
    className="nav"
  >
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pagerLeft"
        name="navigate_before"
        size={24}
      />
    </Button>
    <div
      className="pagesWrap"
    >
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        11
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        12
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        13
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        14
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        15
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        16
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        17
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        18
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        19
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        20
      </Button>
    </div>
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pager"
        name="navigate_next"
        size={24}
      />
    </Button>
  </div>
</div>
`;

exports[`Pager component #render change start and end pages when currentPage <= 6 1`] = `
<div
  className="holdPager"
>
  <div
    className="perpage"
  >
    <Select
      className="select"
      closeOnClickOutside={true}
      defaultValue={15}
      disabled={false}
      height="auto"
      isMobile={false}
      items={Array []}
      onSelect={[Function]}
      open={null}
      position="bottom"
      sortItems={false}
      type="select"
      value={false}
    >
      <SelectOption
        value={10}
      >
        10 por página
      </SelectOption>
      <SelectOption
        value={15}
      >
        15 por página
      </SelectOption>
      <SelectOption
        value={50}
      >
        50 por página
      </SelectOption>
    </Select>
  </div>
  <div
    className="showing"
  >
    60
     - 
    80
     de 
    500
  </div>
  <div
    className="nav"
  >
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pagerLeft"
        name="navigate_before"
        size={24}
      />
    </Button>
    <div
      className="pagesWrap"
    >
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        1
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        2
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        3
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        4
      </Button>
      <Button
        active={false}
        className="btn number isActive"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        5
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        6
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        7
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        8
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        9
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        10
      </Button>
    </div>
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pager"
        name="navigate_next"
        size={24}
      />
    </Button>
  </div>
</div>
`;

exports[`Pager component #render change start and end pages when is a default value 1`] = `
<div
  className="holdPager"
>
  <div
    className="perpage"
  >
    <Select
      className="select"
      closeOnClickOutside={true}
      defaultValue={15}
      disabled={false}
      height="auto"
      isMobile={false}
      items={Array []}
      onSelect={[Function]}
      open={null}
      position="bottom"
      sortItems={false}
      type="select"
      value={false}
    >
      <SelectOption
        value={10}
      >
        10 por página
      </SelectOption>
      <SelectOption
        value={15}
      >
        15 por página
      </SelectOption>
      <SelectOption
        value={50}
      >
        50 por página
      </SelectOption>
    </Select>
  </div>
  <div
    className="showing"
  >
    100
     - 
    120
     de 
    500
  </div>
  <div
    className="nav"
  >
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pagerLeft"
        name="navigate_before"
        size={24}
      />
    </Button>
    <div
      className="pagesWrap"
    >
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        3
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        4
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        5
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        6
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        7
      </Button>
      <Button
        active={false}
        className="btn number isActive"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        8
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        9
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        10
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        11
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        12
      </Button>
    </div>
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pager"
        name="navigate_next"
        size={24}
      />
    </Button>
  </div>
</div>
`;

exports[`Pager component #render render correctly 1`] = `
<div
  className="holdPager"
>
  <div
    className="perpage"
  >
    <Select
      className="select"
      closeOnClickOutside={true}
      defaultValue={15}
      disabled={false}
      height="auto"
      isMobile={false}
      items={Array []}
      onSelect={[Function]}
      open={null}
      position="bottom"
      sortItems={false}
      type="select"
      value={false}
    >
      <SelectOption
        value={10}
      >
        10 por página
      </SelectOption>
      <SelectOption
        value={15}
      >
        15 por página
      </SelectOption>
      <SelectOption
        value={50}
      >
        50 por página
      </SelectOption>
    </Select>
  </div>
  <div
    className="showing"
  >
    20
     - 
    40
     de 
    100
  </div>
  <div
    className="nav"
  >
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pagerLeft"
        name="navigate_before"
        size={24}
      />
    </Button>
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pager"
        name="navigate_next"
        size={24}
      />
    </Button>
  </div>
</div>
`;

exports[`Pager component #render render correctly with hasFirstLast 1`] = `
<div
  className="holdPager"
>
  <div
    className="perpage"
  >
    <Select
      className="select"
      closeOnClickOutside={true}
      defaultValue={15}
      disabled={false}
      height="auto"
      isMobile={false}
      items={Array []}
      onSelect={[Function]}
      open={null}
      position="bottom"
      sortItems={false}
      type="select"
      value={false}
    >
      <SelectOption
        value={10}
      >
        10 por página
      </SelectOption>
      <SelectOption
        value={15}
      >
        15 por página
      </SelectOption>
      <SelectOption
        value={50}
      >
        50 por página
      </SelectOption>
    </Select>
  </div>
  <div
    className="showing"
  >
    20
     - 
    40
     de 
    100
  </div>
  <div
    className="nav"
  >
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pager"
        name="first_page"
        size={24}
      />
    </Button>
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pagerLeft"
        name="navigate_before"
        size={24}
      />
    </Button>
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pager"
        name="navigate_next"
        size={24}
      />
    </Button>
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pager"
        name="last_page"
        size={24}
      />
    </Button>
  </div>
</div>
`;

exports[`Pager component #render render correctly with hasPagination 1`] = `
<div
  className="holdPager"
>
  <div
    className="perpage"
  >
    <Select
      className="select"
      closeOnClickOutside={true}
      defaultValue={15}
      disabled={false}
      height="auto"
      isMobile={false}
      items={Array []}
      onSelect={[Function]}
      open={null}
      position="bottom"
      sortItems={false}
      type="select"
      value={false}
    >
      <SelectOption
        value={10}
      >
        10 por página
      </SelectOption>
      <SelectOption
        value={15}
      >
        15 por página
      </SelectOption>
      <SelectOption
        value={50}
      >
        50 por página
      </SelectOption>
    </Select>
  </div>
  <div
    className="showing"
  >
    20
     - 
    40
     de 
    100
  </div>
  <div
    className="nav"
  >
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pagerLeft"
        name="navigate_before"
        size={24}
      />
    </Button>
    <div
      className="pagesWrap"
    >
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        1
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        2
      </Button>
      <Button
        active={false}
        className="btn number isActive"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        3
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        4
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        5
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        6
      </Button>
      <Button
        active={false}
        className="btn number"
        color="transparent"
        disabled={false}
        onClick={[Function]}
        size="none"
        type="button"
      >
        7
      </Button>
    </div>
    <Button
      active={false}
      className="btn"
      color="transparent"
      disabled={false}
      onClick={[Function]}
      size="none"
      type="button"
    >
      <Icon
        align="middle"
        className="pager"
        name="navigate_next"
        size={24}
      />
    </Button>
  </div>
</div>
`;
